internal/compiler-bridge/src/main/scala-2.12/xsbt/Compat.scala

```diff
@@ -13,9 +13,21 @@ package xsbt
 
 import java.io.PrintWriter
 import xsbti.compile.Output
+import scala.tools.nsc.Global
 import scala.tools.nsc.Settings
 
-abstract class Compat
+abstract class Compat {
+  val global: Global
+  import global._
+
+  /** If given tree contains object tree attachment calls func on tree from attachment. */
+  protected def processOriginalTreeAttachment(in: Tree)(func: Tree => Unit): Unit = {
+    import analyzer._
+    in.attachments.get[OriginalTreeAttachment].foreach { a =>
+      func(a.original)
+    }
```

Comment:
This failed integration test in sbt (https://github.com/sbt/sbt/pull/6591) because `OriginalTreeAttachment` appeared in Scala 2.12.3.

```
[info] Non-compiled module 'compiler-bridge_2.12' for Scala 2.12.1. Compiling...
/var/folders/hg/2602nfrs2958vnshglyl3srw0000gn/T/sbt_f35b067/xsbt/Compat.scala:26: error: not found: type OriginalTreeAttachment
    in.attachments.get[OriginalTreeAttachment].foreach { a =>
                       ^
```

I guess we can drop the support of Scala 2.12.0 - 2.12.2 at some point, but I don't think we can during 1.5.x series. /cc @retronym

Reply:
In the meantime, how about:

```diff
diff --git a/internal/compiler-bridge/src/main/scala-2.12/xsbt/Compat.scala b/internal/compiler-bridge/src/main/scala-2.12/xsbt/Compat.scala
index 13c9d7724..c7d4ac4da 100644
--- a/internal/compiler-bridge/src/main/scala-2.12/xsbt/Compat.scala
+++ b/internal/compiler-bridge/src/main/scala-2.12/xsbt/Compat.scala
@@ -20,11 +20,18 @@ abstract class Compat {
   val global: Global
   import global._
 
+  private lazy val OriginalTreeAttachmentTag = try {
+    reflect.classTag[OriginalTreeAttachment]
+  } catch {
+    case _: LinkageError => null // OriginalTreeAttachment is not in 2.12.{1,2}
+  }
+
   /** If given tree contains object tree attachment calls func on tree from attachment. */
   protected def processOriginalTreeAttachment(in: Tree)(func: Tree => Unit): Unit = {
     import analyzer._
-    in.attachments.get[OriginalTreeAttachment].foreach { a =>
-      func(a.original)
+    OriginalTreeAttachmentTag match {
+      case null =>
+      case tag => in.attachments.get[OriginalTreeAttachment](tag).foreach(a => func(a.original))
     }
   }
 }

```

Reply:
`Compat.scala` is compiled on user's laptop against Scala 2.12.1 so any reference to `OriginalTreeAttachment` would cause compilation error?

This is what I get for example when I set scalaVersion to 2.12.2:

```
[error] /Users/eed3si9n/work/sbt-modules/zinc/internal/compiler-bridge/src/main/scala-2.12/xsbt/Compat.scala:24:22: not found: type OriginalTreeAttachment
[error]     reflect.classTag[OriginalTreeAttachment]
[error]                      ^
```

Reply:
Oh, in that case the lookup of the tag and the call to `.original` needs to be reflective.

Roughly:

```diff
diff --git a/internal/compiler-bridge/src/main/scala-2.12/xsbt/Compat.scala b/internal/compiler-bridge/src/main/scala-2.12/xsbt/Compat.scala
index 13c9d7724..e0f3f007b 100644
--- a/internal/compiler-bridge/src/main/scala-2.12/xsbt/Compat.scala
+++ b/internal/compiler-bridge/src/main/scala-2.12/xsbt/Compat.scala
@@ -23,9 +23,7 @@ abstract class Compat {
   /** If given tree contains object tree attachment calls func on tree from attachment. */
   protected def processOriginalTreeAttachment(in: Tree)(func: Tree => Unit): Unit = {
     import analyzer._
-    in.attachments.get[OriginalTreeAttachment].foreach { a =>
-      func(a.original)
-    }
+    OriginalTreeTraverser.Instance.traverseOriginal(in)(func)
   }
 }
 object Compat {
@@ -34,6 +32,34 @@ object Compat {
 
   // IMain in 2.13 accepts ReplReporter
   def replReporter(settings: Settings, writer: PrintWriter) = writer
+
+  private class OriginalTreeTraverser {
+    def traverseOriginal(t: T)(f: T => Unit): Unit = ()
+  }
+  object OriginalTreeTraverser {
+    private[this] val OriginalTreeClass = 
+      try Class.forName("scala.tools.nsc.typechecker.StdAttachments$OriginalTreeAttachment") catch case _ = > null
+    private[this] val OriginalTreeClassTag = OriginalTreeClass match {
+      case null => null
+      case cls => reflect.ClassTag(cls)
+    }
+    // Compiled into a final static field
+    private[this] val OriginalTree_original = OriginalTreeClass match {
+      case null => null
+      case cls => Class.getMethod("original")
+    }
+    private object Reflective extends OriginalTreeTraverser {
+      def traverseOriginal(t: T)(f: T => Unit): Unit = {
+        t.get(OriginalTreeClassTag) match {
+          case Some(attachment) => f(OriginalTree_original.invoke(attachment))
+          case None =>
+        }
+    }
+    private object NoOp extends OriginalTreeTraverser {
+      def traverseOriginal(t: T)(f: T => Unit): Unit = ()
+    }
+    val Instance = if (OriginalTree_original != null) Reflective else NoOp
+  }
 }
 
 /** Defines compatibility utils for [[ZincCompiler]]. */

```

Reply:
Yea. That makes sense.